v4.9.0 - Refactor

viru185 opened the pull request:
**refactor: drop `pygame` in favor of built-in `PySide6` sound**
  - Removed `pygame` dependency (too heavy just for notifications).
  - Replaced with `QSoundEffect`, which is lightweight and built into `PySide6`.
  - Dropped `pygame.mixer` + threading → Qt handles async playback.
  - Implemented sound playback with `QUrl.fromLocalFile()` and `.play()`.
  - Added `setVolume(0.9)` as a configurable example.
  - Converted notification sound from `.mp3` to `.wav` (only format supported).

**refactor(utils): simplify logger module**
  - Moved `logger` to `src.utils`
  - Removed unnecessary import checks (logger is always available)
  - Replaced `raise` statements with error logging to prevent crashes
  - Use `logger.exception()` in `except` blocks to capture traceback (logged as error)

**style: remove redundant str() in f-strings**
  - Dropped explicit `str()` calls inside f-strings
  - f-strings already call `str()` under the hood

**chore: add type hints for GUI mixins**
  - Added type hints for `FormatTableMixin` (`src.gui.ytsage_gui_format_table`)
  - Added type hints for `VideoInfoMixin` (`src.gui.ytsage_gui_video_info`)
  - Improves autocomplete and type safety in IDEs

**introduce the `ytsage_config_manager.py` module to manage app setting.**
> need to implement this new module
  - Loads settings from a JSON config file (`APP_CONFIG_FILE`).
  - Creates the config file with default values if missing or corrupt.
  - Retrieves, sets, and deletes settings using simple dot-separated keys.
  - Persists updates back to disk automatically.

**Usage**
```python
from src.utils.ytsage_config_manager import ConfigManager

# Load settings (auto-loads if not already loaded)
download_path = ConfigManager.get("download_path")

# Update a value
ConfigManager.set("download_path", "D:/Downloads")

# Retrieve nested value
last_check = ConfigManager.get("cached_versions.ytdlp.last_check")

# Delete a key
ConfigManager.delete("cached_versions.ffmpeg.path")
```

**Refactore: pkg_resources with importlib.metadata.version**
  - UserWarning: pkg_resources is deprecated as an API.
  - See https://setuptools.pypa.io/en/latest/pkg_resources.html.
  - The pkg_resources package is slated for removal as early as 2025-11-30.

## note
- I choose the QSoundEffect as I prefer less dependency and worked perfectly on Windows 11 and kali linux.
- I you still prefer the pyglet, please let me know, I will change it back.

viru185 committed:
fix imports
remove unused imports
use full import path
sort import (1. Standard Library, 2. Third-Party, 3. Local) in alphabetic.

viru185 committed:
- remove: Method 3 from src.core.ytsage_downlader:cleanup_subtitle_file
  - it could delete the subtitle file of other movies if it present in same directory as it scane recursively.

- refactor: migrate from os.path to pathlib.Path for path handling
  - Replaced os.path methods with pathlib.Path to improve readability,
  - avoid repeatation.
  - cross-platform compatibility, and maintain cleaner code.

- improve: enhance code readability
  - Standardized string literals to use double quotes for consistency
  - Removed unnecessary spaces to maintain cleaner formatting
  - Applied code formatting for better readability and maintainability

viru185 committed:
- add: ytsage_constants.py file for one place to store all constants.

- imporve: return type hint for function.

- remove: src/gui/ytsage_gui_dialogs.py file to avoid repetation
  - src/gui/dialogs is renamed to src/gui/ytsage_gui_dialogs for same naming convection. (future import will remains same)
  - use of src/gui/ytsage_gui_dialogs/__init__.py to import the dilogs modules.

- change: variable self.parent to self._parent so it does not overwrite the parent()
  - add type hint checking.

viru185 committed:
- refactor: QMetaObject.invokeMethod to Signal
  - I encounter error with incokeMethod. Could not solve it.
  - So, Changed it to Signal to match app code language.

- implement: the ytsage_constants.py to code
  - remove: unnecessary logic
  - remove: repetitive code logic.

- update: yt-dlp logic for src\gui\ytsage_gui_dialogs\ytsage_dialogs_update:_update_binary
  - yt-dlp update logic will use `yt-dlp -U`

viru185 committed:
refactor: remove unused imports and streamline code formatting across multiple files

viru185 committed:
- **refactor: drop `pygame` in favor of built-in `PySide6` sound**

- Removed `pygame` dependency (too heavy just for notifications).
- Replaced with `QSoundEffect`, which is lightweight and built into `PySide6`.
- Dropped `pygame.mixer` + threading → Qt handles async playback.
- Implemented sound playback with `QUrl.fromLocalFile()` and `.play()`.
- Added `setVolume(0.9)` as a configurable example.
- Converted notification sound from `.mp3` to `.wav` (only format supported).

viru185 committed:
**refactor(utils): simplify logger module**
  - Moved `logger` to `src.utils`
  - Removed unnecessary import checks (logger is always available)
  - Replaced `raise` statements with error logging to prevent crashes
  - Use `logger.exception()` in `except` blocks to capture traceback (logged as error)

**style: remove redundant str() in f-strings**
  - Dropped explicit `str()` calls inside f-strings
  - f-strings already call `str()` under the hood

**chore: add type hints for GUI mixins**
  - Added type hints for `FormatTableMixin` (`src.gui.ytsage_gui_format_table`)
  - Added type hints for `VideoInfoMixin` (`src.gui.ytsage_gui_video_info`)
  - Improves autocomplete and type safety in IDEs

viru185 committed:
- **introduce the `ytsage_config_manager.py` module to manage app setting.**
  - Loads settings from a JSON config file (`APP_CONFIG_FILE`).
  - Creates the config file with default values if missing or corrupt.
  - Retrieves, sets, and deletes settings using simple dot-separated keys.
  - Provides safe error handling with logging instead of raising exceptions.
  - Persists updates back to disk automatically.

- **Usage**
```python
from src.utils.ytsage_config_manager import ConfigManager

download_path = ConfigManager.get("download_path")

ConfigManager.set("download_path", "D:/Downloads")

last_check = ConfigManager.get("cached_versions.ytdlp.last_check")

ConfigManager.delete("cached_versions.ffmpeg.path")
```

viru185 committed:
- Merge branch 'refactor/code-cleanup' into beta
- resolve conflict

viru185 committed:
**Refactore: pkg_resources with importlib.metadata.version**
  - UserWarning: pkg_resources is deprecated as an API.
  - See https://setuptools.pypa.io/en/latest/pkg_resources.html.
  - The pkg_resources package is slated for removal as early as 2025-11-30.

oop7 commented:
Thanks for the updated PR! I've reviewed and tested the changes, and while most improvements look great, I need to request some adjustments:

## Audio Implementation
After testing, `QSoundEffect` actually has a larger footprint than pyglet because it requires `QtMultimedia` and `PySide6.QtNetwork` modules. Combined with the MP3→WAV conversion requirement, pyglet remains the better choice for our notification use case.

## Runtime Issues in Packaged Build
The build process completes successfully, but the packaged executable crashes at runtime with these errors:

**Logger runtime error:**
```
TypeError: Cannot log to objects of type 'NoneType'
```

**Import runtime error:**
```
ModuleNotFoundError: No module named 'yt_dlp'
```

These issues only appear when running the built executable, not during the build process itself. This suggests the logger refactoring and import changes need adjustment for the packaged runtime environment.

**To test this yourself:** Create a tag in your fork to trigger our `.github/workflows/build-windows.yml` pipeline, then download and run the built executable to see these runtime errors.

## What needs to be done:
1. **Revert audio changes** - keep the current pyglet implementation and MP3 files
2. **Fix runtime issues** - ensure changes work when the app is actually executed as a packaged build
3. **Keep the good stuff** - config manager, type hints, and pkg_resources removal are excellent

Please test the actual runtime behavior of packaged builds, not just the build process itself.

Thanks for your patience with this process!

viru185 committed:
**refactore: notification sound**
  - `QSoundEffect` is chnaged back to `pyglet` as per mainter `@oop7` choise.
  - simplify the logic.

viru185 commented:
**refactore: notification sound**
  - `QSoundEffect` is changed back to `pyglet` as per mainter `@oop7` choice.
  - simplify the logic.

viru185 committed:
remove: import check, it should always work.

viru185 committed:
**fix: runtime error**
  - yt_dlp moved from `--excludes` to `--packages` in `build-windows.yml`
  - In frozen build, logger will not log to consol. insted will log to file.
  - In frozen build, `app_dir` is next to `.exe` file.

viru185 commented:
**fix: runtime error**
  - Remove import check, it should always work.
  - `yt_dlp` moved from `--excludes` to `--packages` in `build-windows.yml`
  - In frozen build, logger will not log to console, instead will log to file.
  - In frozen build, `app_dir` is next to `.exe` file.
 
checkout [fix/runtime_issues](https://github.com/viru185/YTSage/releases/tag/v4.8.3) for the build by workflow.

## bugs
1. Selecting new download path resets the config file. can be fixed with new `config manager`.
2. ffmpeg module mess up with `system environment variable`, it removes user added `path`.

viru185 commented:
# very confusing
Can you please explain how you want the `yt-dlp` workflow? here is what I understand. 
- You added `yt-dlp` dependency  to the `requirements.txt` file, imported `yt-dlp` module, used it in the app but did not packed it with the app, then how app going to find the import. of course, it will crash the app. even if you check for for the import, the app will never find the import. It will only work on development. 

## My Suggestion
- Only depend on the imported `yt-dlp` module or remove all together, remove it from the requirements. Here is how it can be handled.
- Check for the `yt-dlp` module install by user with the help of the package manager (`winget, pip, homebrew, apt`), if found use it. otherwise, `yt-dlp` binary will be downloaded to app bin directory. save the `yt-dlp` type (user install or app handled) and location to config file. It can be handled by the `ytsage_config_manager.py`.

> note: we can only handle update app handled binary automatically, we should prompt  to the user to update for user install binary with `os` specific instruction.

## `ffmpeg` critical bug
When ever I try to test the app, the app try to find the system install `ffmpeg` or check for the binary path. If it does not found it, download the v7.1.1 and try to add to system environment variable path instead of using as path.
### To reproduce the bug
> only tested on the Windows 11.
- goal is to app should not be able to find it in environment path. try these command. 
	- `Get-Command <app_name> | Format-List`
	- `where`
	- `Get-Alias | Format-Table -AutoSize`
	- remove system install `ffmpeg` package, remove it from environment variable for user and system, remove the downloaded binary by the app.
- lunch the app and let it handle the `ffmpeg` package installation.
## My general suggestion for handling third party package.
1.  check for user install package (`yt-dlp` and `ffmpeg`)
2. If not found, download it to app binary path. save the path to config file for later reference. also store the if it user install or app managed.

> I will handle these changes, once everything is clarified.

viru185 committed:
Merge remote-tracking branch 'upstream/beta' into beta

viru185 commented:
You’re right that with PyInstaller one-file or AppImage, bundled modules can’t be updated independently. But that’s already how single-file builds work: *any update requires replacing the entire app*. So bundling `yt-dlp` just means updates come through new app releases, which is a normal expectation for desktop users.

For cx\_Freeze folder-based builds, bundling works perfectly and simplifies things:

* No dual logic or subprocess parsing,
* Consistent behavior across dev and frozen builds,
* Better error handling for the GUI.

The current `yt-dlp` auto-update logic would simply be replaced with an **app-level update check**, so users always update the whole package. Given the minimal size impact (\~3 MB), bundling-only still seems the cleanest, most consistent long-term solution.

### Migration Plan (to bundling-only)

1. **Remove** all logic related to `get_yt_dlp_path()`, `check_ytdlp_binary()`, and subprocess calls.
2. **Drop** the `yt-dlp` auto-update mechanism.
3. **Add** an app-level update check (replace current binary updater).
4. **Bundle** `yt-dlp` as a normal dependency in `requirements.txt` and include it in frozen builds.
5. **Ensure** single-file builds simply get new app releases for updates (expected behavior).

viru185 committed:
chang back to checking import for ytdlp

viru185 commented:
For now i have change back to checking import for yt-dlp. so other changes can be merged.

viru185 commented:
## clarification from me

I want to be clear that my main disagreement is with the **hybrid approach**. Maintaining two different code paths (import in dev, binary in frozen builds) adds unnecessary complexity, inconsistent behavior, and confusing requirements. It feels like we get the drawbacks of both methods without the full benefits of either.

I’m fine with us settling on **either** of the two clean approaches:

- **Bundling-only (my preference):** simpler, consistent, better error handling, just update with new app releases.
-  **External-only:** smaller bundle and flexible updates, but at the cost of more code complexity and weaker integration.

Either of those is a clearer long-term strategy than hybrid. For a PySide6 desktop app, I’d still argue bundling-only is the best balance, but my main point is that we should avoid hybrid altogether.

viru185 commented:
I will check it out, please if you don't have any other problem with the current changes, merge it. I have changed it back so that it blend with current branch, no changes to how app integrate yt-dlp package.

oop7 committed:
Bump version to 4.8.1

Update version references from 4.8.0b to 4.8.1 in __init__.py, main app, and About dialog to reflect the new release.

oop7 committed:
Update asset paths in build workflows

Changed asset inclusion and screenshot removal paths from 'assets' to 'lib/assets' in Linux, macOS, and Windows build workflows to reflect new directory structure and ensure screenshots are excluded from packaged builds.

oop7 committed:
Bump version to 4.8.2

Update version references from 4.8.1 to 4.8.2 in source files and documentation to prepare for a new patch release.

oop7 committed:
Fix asset include path in Windows build workflow

Corrects the syntax for including asset files in the build-windows.yml workflow by changing 'assets,lib/assets' to 'assets=lib/assets'. This ensures assets are properly mapped during the build process.

oop7 committed:
Update release tag examples in CI/CD README

Changed the example git tag commands from v4.8.0 and v4.8.2 to v4.8.1 for consistency in the CI/CD documentation.

oop7 committed:
Fix include-files mapping in Windows build workflow

Changed the cx_Freeze --include-files argument from '=' to ':' for source:destination mapping in build-windows.yml. This resolves an issue where '=' was treated as a literal path, ensuring assets are correctly copied to the destination directory.

oop7 committed:
Remove redundant comments in build-windows workflow

Deleted comments explaining the colon usage for source:destination mapping in the cx_Freeze CLI, as the mapping is already clear from the context.

oop7 committed:
Refactor Windows build to use cx_Freeze setup script

Replaces direct cx_Freeze CLI calls with dynamically generated setup scripts for both standard and FFmpeg builds. This improves maintainability and flexibility of build configuration in the GitHub Actions workflow.

oop7 committed:
Bump version to 4.8.3

Updated version references from 4.8.2 to 4.8.3 in __init__.py, main app, and About dialog to reflect the new release.

viru185 committed:
revert(build): move yt_dlp from --packages to --excludes in build-windows.yml

viru185 committed:
Merge remote-tracking branch 'upstream/beta' into beta

viru185 committed:
Merge remote-tracking branch 'upstream/main' into beta
